loaders.py

```diff
@@ -68,6 +70,70 @@ def __getitem__(self, index):
             return Xi
 
 
+class MetadataImageSegmentationDatasetDistances(Dataset):
+    def __init__(self, X, y, train_mode,
+                 image_transform, image_augment_with_target,
+                 mask_transform, image_augment,
+                 distance_matrix_transform):
+        super().__init__()
+        self.X = X
+        if y is not None:
+            self.y = y
+        else:
+            self.y = None
+
+        self.train_mode = train_mode
+        self.image_transform = image_transform
+        self.mask_transform = mask_transform
+        self.distance_matrix_transform = distance_matrix_transform
+        self.image_augment = image_augment
+        self.image_augment_with_target = image_augment_with_target
+
+    def load_image(self, img_filepath):
+        image = Image.open(img_filepath, 'r')
+        return image.convert('RGB')
+
+    def load_distances(selfself, dist_filepath):
+        return joblib.load(dist_filepath)
+
+    def __len__(self):
+        return self.X.shape[0]
+
+    def __getitem__(self, index):
+        img_filepath = self.X[index]
+        Xi = self.load_image(img_filepath)
+
+        if self.y is not None:
+            mask_filepath = self.y[index]
+            Mi = self.load_image(mask_filepath)
+            distance_filepath = mask_filepath.replace("/masks/", "/distances/")[:-4]
+            Di = self.load_distances(distance_filepath)
+            Di = np.sum(Di, axis=2).astype(np.uint8)  # TODO: remove it when Di will be sum of distances to 2 closest objects
+
+            if self.train_mode and self.image_augment_with_target is not None:
+                Xi, Mi = from_pil(Xi, Mi)
+                Xi, Mi, Di = self.image_augment_with_target(Xi, Mi, Di)
+                if self.image_augment is not None:
+                    Xi = self.image_augment(Xi)
+                Xi, Mi, Di = to_pil(Xi, Mi, Di)
+
+            if not self.train_mode:
+                Di = to_pil(Di)
```

Comment:
@apyskir why do we need Di in the inference mode?

Reply:
@jakubczakon actually we don't. But look here https://github.com/minerva-ml/open-solution-mapping-challenge/blob/dev/loaders.py#L97 This suggests, that we need mask in inference mode, but we don't.
Am I right, that when `self.y!=None` then `self.train_mode==True`? Maybe these parameters are equivalent, but sometimes first is used, sometimes the other.

Reply:
@apyskir I see train_mode with validation data still needs to perform transformations on target (not augmentations). That is why this weird situation arises. Nvmd just clumsy coding on my part. I will add an issue